### PR TITLE
http-api-origin: implement customAuth (authorization done with an ext…

### PR DESCRIPTION
**Pull request by anton-107**
…ra http call)

map-transformer: implement "mergeArrays" method

### CODE REVIEW

File: packages/@datapull/http-api-origin/index.js [hunk 1]
```diff
@@ -26,10 +26,85 @@ class HttpApiOrigin {
     };
   }
 
+  async customAuth(config) {
+    return new Promise((resolve, reject) => {
+      if (config.auth !== 'custom') {
+        return resolve();
```
Comment (mishok13) [inline packages/@datapull/http-api-origin/index.js:32]: Did you mean `reject()` here?

File: packages/@datapull/http-api-origin/index.js [hunk 2]
```diff
@@ -26,10 +26,85 @@ class HttpApiOrigin {
...
+
+        let error;
+        if (statusCode !== 200) {
+          error = `Request Failed. Status Code: ${statusCode}`;
+        }
+
+        if (error) {
+          res.resume(); // consume response data to free up memory
+          return reject(error);
+        }
```
Comment (mishok13) [inline packages/@datapull/http-api-origin/index.js]: Isn't the whole block identical to 
```javascript
if (statusCode !== 200) {
  res.resume();
  return reject(`Request Failed. Status Code: ${statusCode}`);
}
```

Also, while we're at it, I think recommended way of rejecting a promise is by passing an `Error` object not just a string.

File: packages/@datapull/http-api-origin/index.js [hunk 3]
```diff
@@ -26,10 +26,85 @@ class HttpApiOrigin {
...
+            if (config.customAuth.format === 'json') {
+              parsedData = JSON.parse(rawData);
+            }
+
+            if (config.customAuth.secretKey) {
+              parsedData = parsedData[config.customAuth.secretKey];
+            }
+
+            if (!parsedData) {
+              return reject("Could not get credentials (custom auth)");
```
Comment (mishok13) [inline packages/@datapull/http-api-origin/index.js]: `reject(new Error(...` ;)

File: packages/@datapull/http-api-origin/index.js [hunk 4]
```diff
@@ -48,6 +123,14 @@ class HttpApiOrigin {
         options.auth = `${config.username}:${config.password}`;
       }
 
+      if (this.config.headers) {
+        options.headers = {};
+        Object.keys(this.config.headers).forEach(k => {
```
Comment (mishok13) [inline packages/@datapull/http-api-origin/index.js]: My personal preference is that we use `for .. of` for loops instead.

File: packages/@datapull/map-transformer/index.js [hunk 5]
```diff
@@ -14,7 +15,11 @@ class MapTransformer {
           r.push(transformedItem);
         });
       });
+      return Promise.resolve(r);
+    }
 
+    if (this.config === 'mergeArrays') {
+      r = [].concat.apply([], data);
```
Comment (mishok13) [inline packages/@datapull/map-transformer/index.js]: Isn't this identical to `[].concat(...data)`? Also, why assign to `r` if you're immediately returning it in the next line?

**Review by mishok13 — CHANGES_REQUESTED** [review]

File: packages/@datapull/http-api-origin/index.js (hunk 1 above)
Comment (anton-107) [reply packages/@datapull/http-api-origin/index.js:32]: `resolve()` is correct here - it means no "custom" auth is needed, we can proceed to an actual request

File: packages/@datapull/http-api-origin/index.js (hunk 2 above)
Comment (anton-107) [reply packages/@datapull/http-api-origin/index.js]: fixed

File: packages/@datapull/http-api-origin/index.js (hunk 3 above)
Comment (anton-107) [reply packages/@datapull/http-api-origin/index.js]: fixed

File: packages/@datapull/http-api-origin/index.js (hunk 4 above)
Comment (anton-107) [reply packages/@datapull/http-api-origin/index.js]: fixed

File: packages/@datapull/map-transformer/index.js (hunk 5 above)
Comment (anton-107) [reply packages/@datapull/map-transformer/index.js]: fixed

**Review by mishok13 — APPROVED** [review]

**Pull request merged**